pkgs/fc/agent/fc/util/slurm.py

```diff
@@ -0,0 +1,743 @@
+            raise NodeStateError(state, flags)
+        else:
+            log.info(
+                "down-already-reached",
+                _replace_msg="{node} is already in {state} state.",
+                state=state,
+            )
+            return
+
+    state_change_down = {
+        "node_names": node_name,
+        "node_state": pyslurm.NODE_STATE_DOWN,
+        "reason": reason,
+    }
+    result = update_nodes(state_change_down)
+    log.debug("node-update-result", result=result)
+    log.info(
+        "down-finished",
+        _replace_msg="{node} is set to DOWN.",
+    )
+
+
+class ReadyPreCheckResult(NamedTuple):
+    state: str
+    flags: list[str]
+    action: bool
+
+
+def run_ready_pre_checks(
+    log,
+    node_name,
+    strict_state_check,
+    reason_must_match,
+    skip_in_maintenance,
+    directory,
+):
+    log = log.bind(node=node_name)
+    node_info = get_node_info(node_name)
+    state, *flags = node_info["state"].split("+")
+    log.debug("ready-pre-node-state", state=state, flags=flags)
+
+    if state in ("ALLOCATED", "IDLE", "MIXED"):
+        if strict_state_check:
+            log.error("ready-state-error", state=state, flags=flags)
+            raise NodeStateError(state, flags)
+        else:
+            log.info(
+                "ready-already-reached",
+                _replace_msg=(
+                    "{node} is already in a ready state ({state}). No change."
+                ),
+                state=state,
+            )
+            return ReadyPreCheckResult(state, flags, action=False)
+
+    if state in ("ALLOCATED*", "IDLE*", "MIXED*"):
+        if strict_state_check:
+            log.error("ready-state-error", state=state, flags=flags)
+            raise NodeStateError(state, flags)
+        else:
+            log.warn(
+                "ready-already-reached-unresponsive",
+                _replace_msg=(
+                    "{node} is already in a ready state ({state}) but not "
+                    "responding at the moment. No change."
+                ),
+                state=state,
+            )
+            return ReadyPreCheckResult(state, flags, action=False)
+
+    if reason_must_match and reason_must_match not in node_info["reason"]:
+        log.info(
+            "ready-pre-reason-not-matched",
+            _replace_msg=(
+                "{node} cannot be ready because the reason '{reason}'"
+                "does not contain the expected string {expected}"
+            ),
+            expected=reason_must_match,
+            reason=node_info["reason"],
+        )
+        return ReadyPreCheckResult(state, flags, action=False)
+
+    if skip_in_maintenance and not is_node_in_service(directory, node_name):
+        log.info(
+            "ready-pre-not-in-service",
+            node=node_name,
+            _replace_msg="{node} is still in maintenance, skipping.",
+        )
+        return ReadyPreCheckResult(state, flags, action=False)
+
+    if "*" in state:
```

Comment:
In an object model world, this could also be made more expressive. Either through a state check API or having states be smart objects, too.

```
if node.state.node_not_responding:
```